File: Semester 5/PPD/proiect/examples/distributed_example.py
"""
Example demonstrating the distributed MPI n-body simulation.

To run this example:
    mpirun -np 4 python examples/distributed_example.py
"""

from shared.utils import generate_random_bodies
from distributed.simulation import DistributedNBodySimulation


def main():
    """Run a distributed n-body simulation."""
    from mpi4py import MPI
    
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    
    if rank == 0:
        print("=" * 60)
        print("Distributed N-Body Simulation (MPI)")
        print("=" * 60)
        print(f"Running on {comm.Get_size()} processes")
        
        # Generate bodies (only on rank 0)
        print("\nGenerating bodies...")
        bodies = generate_random_bodies(n=20, 
                                       mass_range=(1e20, 1e25),
                                       position_range=(-1e6, 1e6),
                                       velocity_range=(-1e3, 1e3))
        print(f"Created {len(bodies)} bodies")
    else:
        bodies = None
    
    # Create distributed simulation
    simulation = DistributedNBodySimulation(bodies)
    
    if rank == 0:
        print(f"\nProcess {rank}: Managing bodies {simulation.start_idx} to {simulation.end_idx}")
        print("Running simulation...")
    
    # Run simulation
    simulation.run(num_steps=100, dt=1.0, verbose=(rank == 0))
    
    # Print statistics (only on rank 0)
    if rank == 0:
        print("\n" + "=" * 60)
        print("Simulation Statistics")
        print("=" * 60)
        stats = simulation.get_statistics()
        for key, value in stats.items():
            if key != 'center_of_mass':
                print(f"{key}: {value}")
        print(f"center_of_mass: ({stats['center_of_mass'][0]:.2e}, "
              f"{stats['center_of_mass'][1]:.2e}, {stats['center_of_mass'][2]:.2e})")


if __name__ == "__main__":
    main()

File: Semester 5/PPD/proiect/cpp/examples/distributed_example.cpp
#include "distributed_simulation.h"
#include "utils.h"
#include <iostream>
#include <iomanip>
#include <mpi.h>

int main(int argc, char* argv[]) {
    MPI_Init(&argc, &argv);
    
    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    
    if (rank == 0) {
        std::cout << "============================================================" << std::endl;
        std::cout << "Distributed N-Body Simulation (MPI)" << std::endl;
        std::cout << "============================================================" << std::endl;
        std::cout << "Running on " << size << " processes" << std::endl;
        
        // Generate bodies (only on rank 0)
        std::cout << "\nGenerating bodies..." << std::endl;
        auto bodies = generateRandomBodies(20, 1e20, 1e25, -1e6, 1e6, -1e3, 1e3);
        std::cout << "Created " << bodies.size() << " bodies" << std::endl;
        
        // Create distributed simulation
        DistributedNBodySimulation simulation(bodies);
        
        std::cout << "\nRunning simulation..." << std::endl;
        simulation.run(100, 1.0, true);
        
        // Print statistics
        std::cout << "\n============================================================" << std::endl;
        std::cout << "Simulation Statistics" << std::endl;
        std::cout << "============================================================" << std::endl;
        auto stats = simulation.getStatistics();
        std::cout << "num_bodies: " << stats.num_bodies << std::endl;
        std::cout << "num_processes: " << stats.num_processes << std::endl;
        std::cout << "time_elapsed: " << std::scientific << std::setprecision(2) 
                  << stats.time_elapsed << std::endl;
        std::cout << "step_count: " << stats.step_count << std::endl;
        std::cout << "total_energy: " << stats.total_energy << std::endl;
        std::cout << "center_of_mass: (" << stats.center_of_mass[0] << ", "
                  << stats.center_of_mass[1] << ", " << stats.center_of_mass[2] << ")" << std::endl;
    } else {
        // Other ranks: create simulation (bodies will be received in constructor)
        std::vector<Body> empty_bodies;  // Will be populated by constructor
        DistributedNBodySimulation simulation(empty_bodies);
        simulation.run(100, 1.0, false);
    }
    
    MPI_Finalize();
    return 0;
}

